aten/src/ATen/core/ivalue.cpp

```diff
@@ -630,6 +631,29 @@ getClassConverter() {
   return classConverter;
 }
 
+void ivalue::Future::addCallback(
+    std::function<void(void)> callback,
+    bool propagateTLSState) {
+  std::function<void(void)> cb;
+  if (propagateTLSState) {
```

Comment:
I'm fine with this approach as-is. But before we land this, I wanted to throw out an alternative to see if we considered it.

In the spirit of unix-y composition of small components, instead of
  `future->addCallback(cb, true /*propagateTLSState*/);`
add optional wrapper like the following:
  `future->addCallback(wrapPropagateTLSState(cb));`

The impl would be basically cutting and pasting the bits we already have, and wouldn't need to touch addCallback(), anymore, just add a short utility function like:

  std::function<void(void)> wrapPropagateTLSState(std::function<void(void)> cb) {
    at::ThreadLocalState tls_state;
    return [tls_state = std::move(tls_state), callback = std::move(callback)] {
        at::ThreadLocalStateGuard g(tls_state);
        callback();
      };
  }

The advantage would mostly be a simpler base impl with no implicit args, and perhaps more maintainability if we ever want to add another arg to the `addCallback` function (hopefully that could be done via composition as well)

Reply:
This sounds good to me, as we won't have to modify addCallback and then() at all.